### PR TITLE
Add coalesce function

### PR DESCRIPTION
**Pull request by rtaket**

### CODE REVIEW

File: XForm/XForm.Test/Functions/String/CoalesceFunctionTests.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        [TestMethod]
+        public void Coalesce()
+        {
+            // Arrange
+
+            String8[] transformedArray = null;
+            bool[] nullArray = null;
+
+            // 1 a !   =   1
```
Comment (vScottLouvau) [inline XForm/XForm.Test/Functions/String/CoalesceFunctionTests.cs]: It wasn't clear to me that Coalesce takes the first non-null/empty value. A comment here would be helpful.

File: XForm/XForm/Functions/String/Coalesce.cs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using XForm.Commands;
+using XForm.Data;
+using XForm.Types;
+
+namespace XForm.Functions.String
+{
+    internal class CoalesceBuilder : IFunctionBuilder
```
Comment (vScottLouvau) [inline XForm/XForm/Functions/String/Coalesce.cs]: Class comment for the behavior of Coalesce?

File: XForm/XForm/Functions/String/Coalesce.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            // Coalesce only works on String8 columns
+            if (columns.Any(column => column.ColumnDetails.Type != typeof(String8)))
+            {
+                throw new ArgumentOutOfRangeException("Every input columns for Coalesce must be of type String8.");
+            }
+
+            return new CoalesceColumn(columns);
+        }
+
+        public ColumnDetails ColumnDetails
```
Comment (vScottLouvau) [inline XForm/XForm/Functions/String/Coalesce.cs]: Move auto property up by Columns?

File: XForm/XForm/Functions/String/Coalesce.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+            ColumnDetails = new ColumnDetails("Coalesce", typeof(String8), true);
+        }
+
+        public static IDataBatchColumn Build(IDataBatchEnumerator source, IEnumerable<IDataBatchColumn> columns)
+        {
+            // Coalesce only works on String8 columns
+            if (columns.Any(column => column.ColumnDetails.Type != typeof(String8)))
+            {
+                throw new ArgumentOutOfRangeException("Every input columns for Coalesce must be of type String8.");
```
Comment (vScottLouvau) [inline XForm/XForm/Functions/String/Coalesce.cs]: ArgumentException

File: XForm/XForm/Functions/String/Coalesce.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        {
+            Allocator.AllocateToSize(ref transformedArray, batches[0].Count);
+            Allocator.AllocateToSize(ref nullArray, batches[0].Count);
+
+            for (int i = 0; i < batches[0].Count; ++i)
+            {
+                foreach (var column in batches)
+                {
+                    // Choose the first non-null/empty value
+                    if ((column.IsNull != null && !column.IsNull[i]) || (String8)column.Array.GetValue(i) != String8.Empty)
```
Comment (vScottLouvau) [inline XForm/XForm/Functions/String/Coalesce.cs]: You're not looking up the index. You need int Index = column.IndexOf(i). The real index is the index you should use into both the null array and the value array.

File: XForm/XForm/Functions/String/Coalesce.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                    {
+                        transformedArray[i] = (String8)column.Array.GetValue(i);
+                        break;
+                    }
+                }
+
+                // Update the null array if the value is null/empty
+                if (transformedArray[i] == String8.Empty)
+                {
+                    nullArray[i] = true;
```
Comment (vScottLouvau) [inline XForm/XForm/Functions/String/Coalesce.cs]: "Fast path" code can't run when there are nulls. Can you track whether you've set any nulls and not pass the nullArray in the result block if you never set anything to true?

Comment (vScottLouvau) [reply XForm/XForm/Functions/String/Coalesce.cs]: Also, I think you need to clear nullArray and transformedArray before your loop over the batches. It looks like they'll keep the values from a prior batch iteration.

**Comment by jeff-bacon** [comment]
Coalesce has been implemented in another commit.

**Pull request closed without merging**